docs/src/adr/005adr-json.md

```diff
@@ -0,0 +1,121 @@
+
+## 2. Disambiguation field
+
+Every serialization will contain a disambiguation field, named `kind`. This field holds the name of the class being serialized. For example, the serialization of a `NameEx` will have the shape
+```
+{
+  kind: "NameEx"
+  ...
+}
+```
+
+## 3. Serializing typed entities
+
+Serializations of typed entities will have an additional field named `type`, containing the type of the expression (see [ADR-002][], [ADR-004][] for a description of our type system and the syntax for types-as-string-annotations respectively). For example, the integer literal `1` is represented by a `ValEx`, which has type `Int` and is serialized as follows:
+```
+{
+  kind: "NameEx"
+  type: "Int"
+  ...
+}
+```
+
+## 4. Serialization rules
+
+The goal of the serialization is for the JSON structure to mimic the internal representation as closely as possible, for ease of deserialization. 
+Concretely, whenever a class declares a field `fld: T`, its serialization also contains a field named `fld`, containing the serialization of the field value.
+For example, if `TlaConstDecl` declares a `name: String` field, its JSON serialization will have a `name` field as well, containing the name string.
+
+If a class field has the type `Traversable[T]`, for some `T`, the corresponding JSON entry is a list containing serializations of the individual arguments. For example, `OperEx` is variadic and declares `args: TlaEx*`, so its serialization has an `args` field containing a (possibly empty) list.
+
+As a running example, take the expression `1 + 1`, represented with the correct type annotations as 
+```
+OperEx( 
+  oper = TlaArithOper.plus, 
+  args = Seq( 
+    ValEx( TlaInt( 1 ) )( typeTag = Typed( IntT1() ) ), 
+    ValEx( TlaInt( 1 ) )( typeTag = Typed( IntT1() ) )
+    ) 
+  )( typeTag = Typed( IntT1() ) )
+```
+
+Since both sub-expressions, the literals `1`, are identical, their serializations are equal as well:
+```
+{
+  kind: "ValEx",
+  type: "Int",
+  value: {
+    kind: "TlaInt",
+    value: 1
+  }
+}
+```
+
+Note that in the above example, the outer `value` field contains a JSON serialization of `TlaInt(1)`, and not just the JSON literal `1`. This is because of deserialization:
+Apalache encodes its TLA+ integers as `BigInt`, which means that it permits values, for which `.isValidInt` does not hold. If we chose to encode values directly as JSON numbers, we would need a) a sensible encoding of `BigInt` values, which are not valid integers and b) a way to distinguish both variants of `BigInt`, as well as decimals, when deserializing (since JSON is not typed).
+Alternatively, we could encode all values as strings, which would be similarly indistinguishable when deserializing. Importantly, the `type` field is not guaranteed to contain a hint, as it could be `Untyped`.
+For the above reasons, we choose to instead serialize `TlaValue` as a JSON object, which is more verbose, but trivial to deserialize. It has the following shape
```

Comment:
Can you add a bit more structure to this text? For instance, describe the general rules in Section "1. General rules" and the rules for Int in Section "2. Rules for Int". Then it will be clear that Int is a special case, and the rest follows the general principle.